File: ios/RaiseTimer/Domain/PayoutCalculator.swift
import Foundation

enum PayoutCalculator {
    struct Payout: Identifiable, Hashable {
        var id: Int { place }
        var place: Int
        var amount: Int
    }

    static func calculate(_ state: TournamentState) -> [Payout] {
        let percents = state.config.payoutPercents
        guard !percents.isEmpty else { return [] }

        let pool = state.totalPrizePool
        if pool <= 0 {
            return percents.indices.map { Payout(place: $0 + 1, amount: 0) }
        }

        let sum: Int = max(percents.reduce(0, +), 1)
        let poolLong: Int64 = Int64(pool)
        let sumLong: Int64 = Int64(sum)
        var rawAmounts: [Int] = []
        rawAmounts.reserveCapacity(percents.count)
        for pct in percents {
            let amount: Int64 = (poolLong * Int64(pct)) / sumLong
            rawAmounts.append(Int(amount))
        }
        let distributed: Int = rawAmounts.reduce(0, +)
        let remainder: Int = pool - distributed

        return rawAmounts.enumerated().map { index, amount in
            let adjusted = index == 0 ? amount + remainder : amount
            return Payout(place: index + 1, amount: adjusted)
        }
    }
}

File: android/app/src/main/java/com/jiny/raisetimer/domain/PayoutCalculator.kt
package com.jiny.raisetimer.domain

import com.jiny.raisetimer.domain.model.TournamentState

object PayoutCalculator {

    data class Payout(val place: Int, val amount: Long)

    /**
     * Splits the total prize pool across [TournamentConfig.payoutPercents]. Percentages are
     * normalized if they do not sum to 100, and any remainder from integer division is
     * added to the first place.
     */
    fun calculate(state: TournamentState): List<Payout> {
        val percents = state.config.payoutPercents
        if (percents.isEmpty()) return emptyList()

        val pool = state.totalPrizePool
        if (pool <= 0) return percents.indices.map { Payout(it + 1, 0L) }

        val sum = percents.sum().takeIf { it > 0 } ?: 100
        val rawAmounts = percents.map { pct -> pool * pct / sum }
        val distributed = rawAmounts.sum()
        val remainder = pool - distributed

        return rawAmounts.mapIndexed { index, amount ->
            val adjusted = if (index == 0) amount + remainder else amount
            Payout(place = index + 1, amount = adjusted)
        }
    }
}
